manifest/dynamic_manifest.go

```diff
@@ -0,0 +1,226 @@
+	}
+	go m.handleIncomingManifests(ctx, errCh)
+	m.handleApplyManifest(ctx, errCh)
+}
+
+func (m *DynamicManifestProvider) handleApplyManifest(ctx context.Context, errCh chan error) {
+	// add a timer for EC period
+	ticker := time.NewTicker(ManifestCheckTick)
+	defer ticker.Stop()
+
+	for {
+		select {
+		case <-ticker.C:
+			m.lk.Lock()
+			if m.nextManifest != nil {
+				ts, err := m.ec.GetHead(ctx)
+				if err != nil {
+					log.Errorf("error fetching chain head: %+v", err)
+					m.lk.Unlock()
+					continue
+				}
+
+				// if the upgrade epoch is reached or already passed.
+				if ts.Epoch() >= m.nextManifest.BootstrapEpoch {
+					log.Debugf("reached bootstrap epoch, triggering manifest change: %d", ts.Epoch())
+					// update the current manifest
+					prevManifest := m.manifest
+					m.manifest = *m.nextManifest
+					nn := m.networkNameOnChange()
+					m.manifest.NetworkName = nn
+					m.nextManifest = nil
+					// trigger manifest change callback.
+					go m.onManifestChange(ctx, prevManifest, errCh)
+					m.lk.Unlock()
+					continue
+				}
+			}
+			m.lk.Unlock()
+		case <-ctx.Done():
+			return
+		}
+	}
+}
+
+// listen to manifests being broadcast through the network.
+func (m *DynamicManifestProvider) handleIncomingManifests(ctx context.Context, errCh chan error) {
+	if err := m.setupManifestPubsub(); err != nil {
+		errCh <- xerrors.Errorf("setting up pubsub: %w", err)
+		return
+	}
+
+	manifestSub, err := m.manifestTopic.Subscribe()
+	if err != nil {
+		errCh <- xerrors.Errorf("subscribing to topic: %w", err)
+		return
+	}
+
+loop:
+	for {
+		select {
+		case <-ctx.Done():
+			break loop
+
+		default:
+			var msg *pubsub.Message
+			msg, err = manifestSub.Next(ctx)
+			if err != nil {
+				if ctx.Err() != nil {
+					break
+				}
+				log.Errorf("manifestPubsub subscription.Next() returned an error: %+v", err)
+				break
+			}
+			manifest, ok := msg.ValidatorData.(*Manifest)
+			if !ok {
+				log.Errorf("invalid manifestValidatorData: %+v", msg.ValidatorData)
+				continue
+			}
+
+			m.acceptNextManifest(manifest)
+		}
+	}
+
+	manifestSub.Cancel()
+	if err := m.teardownManifestPubsub(); err != nil {
+		errCh <- xerrors.Errorf("shutting down manifest pubsub: %w", err)
+	}
+}
+
+func (m *DynamicManifestProvider) teardownPubsub(topic *pubsub.Topic, topicName string) error {
+	return multierr.Combine(
+		m.pubsub.UnregisterTopicValidator(topicName),
+		topic.Close(),
+	)
+}
+
+func (m *DynamicManifestProvider) teardownManifestPubsub() error {
+	return m.teardownPubsub(m.manifestTopic, ManifestPubSubTopicName)
+}
+
+// Checks if we should accept the manifest that we received through pubsub
+// and sets nextManifest if it is the case
+func (m *DynamicManifestProvider) acceptNextManifest(manifest *Manifest) {
+	m.lk.Lock()
+	defer m.lk.Unlock()
+
+	if manifest.Sequence <= m.manifest.Sequence {
+		return
+	}
+	// TODO: Any additional logic here to determine what manifests to accept or not?
```

Comment:
Double checking that the manifest signature/sender is confirmed to be the passive test control server before we hit this.